polynote-kernel/src/main/scala/polynote/kernel/dependency/DependencyFetcher.scala

```diff
@@ -208,6 +185,76 @@ class CoursierFetcher extends DependencyFetcher[IO] {
     run(resolution.process, 0)
   }
 
+  def splitDependencies(deps: List[DependencyConfigs]): (List[URI], List[DependencyConfigs]) = {
+    val (urlList, dependencies) = deps.flatMap { dep =>
+      dep.toList.collect {
+        case (k, v) =>
+          val (dependencies, urls) = v.map { s =>
+            Either.fromTry {
+              Try(new URL(s)) // use URL because its parser is strict, even though we actually want URI
+                .map(_.toURI)
+                // we might still know how to handle the protocol even if URL doesn't
+                .recoverWith {
+                  case t: MalformedURLException if t.getMessage.contains("unknown protocol") => Try(new URI(s))
+                }
+            }.leftMap(_ => s) // on the left, strings are dependency coordinates
+          }.separate
+
+          (urls, TinyMap(k -> TinyList(dependencies)))
+      }
+    }.unzip
+    (urlList.flatten, dependencies)
+  }
+
+  def fetchUrls(uris: List[URI], statusUpdates: Publish[IO, KernelStatusUpdate], chunkSize: Int = 8192)(implicit ctx: ExecutionContext): List[(String, IO[File])] = {
+
+    def withProgress(s: fs2.Stream[IO, Byte], uri: String, fileSize: Long): fs2.Stream[IO, Byte] = {
+      val size = if (fileSize > 0) fileSize else 300 * 1024 * 1024 // this is roughly the size of the largest uber jars we have
+
+      // is there a better way to do this? seems more convoluted than necessary but I couldn't quite get there...
+      // in practice all this gobbledygook doesn't seem to slow things down though, so that's good.
```

Comment:
Yeah this is actually a pretty efficient way to do it... imagine if you were doing `mapAccumulate` on the bytes and counting them individually! Most things in fs2 are surprisingly performant (except when they're not)

Reply:
Haha in fact one of the iterations I was playing with I did that and it ground to a halt.